teps/0002-custom-tasks.md

```diff
@@ -0,0 +1,392 @@
+```
+
+### Pipeline Integration
+
+Enabling `Run`s by themselves are not terribly compelling. Their power comes from being specified in `Pipeline`s, and executed during `PipelineRun`s.
+
+Under this proposal, a user can define a Pipeline that invokes a Custom Task, specified similar to how `Task`s are specified today:
+
+```
+apiVersion: tekton.dev/v1beta1
+kind: Pipeline
+metadata:
+  name: example-pipeline
+spec:
+  tasks:
+  ...
+  - name: example-task
+    taskRef:
+      apiVersion: example.dev/v0
+      kind: Example
+      name: my-example
+```
+
+When a `PipelineRun` is created referencing this `Pipeline` definition, when `example-task` is next to execute, Tekton's `PipelineRun` controller will create a `Run` referencing the `Example` object, at which point it the Custom Task author's controller will be notified and it will be that controller's responsibility to take some action and eventually update the `Run`'s `.status` to signal completion.
+
+The `PipelineRun` controller will watch `Run` objects it's created, and take appropriate action when they report success or failure.
+
+### Initial Update Timeout
+
+It's possible that, though the Custom Task author has defined their CRD type, there's no controller watching for `Run`s of that type and updating their statuses as expected (e.g., it's crash-looping, it has been uninstalled, or it was never provided).
+
+In this case, a `PipelineRun` that depends on that type may wait uselessly for the execution to complete, until its configured timeout, which might be hours later. To save users' time and fail fast, the `PipelineRun`  controller will enforce a short timeout for initial updates to `Run` objects. If a `Run` hasn't been updated to the condition `.status.conditions[@type=Succeeded]=Unknown` after a certain (configurable) amount of time, say 30 seconds, then the `PipelineRun`  controller should fail the `PipelineRun` with a descriptive error message. This simple update indicates that the execution is running, and at least something is consuming new `Run` creations and updating statuses.
+
+### Parameter Passing
+
+Custom Task authors should support parameter passing, by supporting a `.spec.params` field (of type [`[]Param`](https://godoc.org/github.com/tektoncd/pipeline/pkg/apis/pipeline/v1beta1#Param)), and by resolving any `$(params.foo)` placeholders in the CRD type when a `Run` of that type is first reconciled -- this functionality should be implemented by a Go package provided by Tekton, which should be the same one that Tekton itself uses when resolving placeholders in `TaskRun`s and `PipelineRun`s.
+
+This Pipeline pipes its input param `"pl-wait-duration"` to the `example-task`:
+
+```
+apiVersion: tekton.dev/v1beta1
+kind: Pipeline
+metadata:
+  name: example
+spec:
+  params:
+  - name: pipeline-param
+    default: hello
+  tasks:
+  - name: example-task
+    taskRef:
+      apiVersion: example.dev/v0
+      kind: Example
+      name: my-example
```

Comment:
@ImJasonH and i were talking about a situation where there _is_ no type to refer to, for example say i want to make a custom controller that evaluates CEL expressions, but I don't need a CEL CRD type to refer to - all I really need is for a run to be created with a param with the expression (not saying this is the final design, just a possibility for this feature or any other). in that case i need to indicate in the Run what type of thing we're running so my controller can filter for it, but there is no instance to refer to

@ImJasonH suggested we could do this by making "name" an optional field

Reply:
@ImJasonH i tried to think of syntax tweaks and i have nothing super compelling but here are my thoughts:

1. it's interesting in the definition of an instance of a type, that apiversion and kind are at a different level than name, which is a field of metadata, maybe we could separate them also somehow:

```
apiVersion: tekton.dev/v1beta1
kind: ClusterTask
metadata:
  name: clustertask-v1beta1
```

2. instead of overloading taskRef, adding a new section called `run` which is used when a Run is supposed to be created, makes me less confused. but it pretty much has the same fields as `taskRef` and defeats the whole purpose of including `kind`in there in the first place

3. instead of a `ref` section in `Run` maybe these values could be top level fields

anyway none of these ideas are great, feel free to ignore them

Reply:
Our [`TaskRef`](https://godoc.org/github.com/tektoncd/pipeline/pkg/apis/pipeline/v1beta1#TaskRef) type is based on [`corev1.ObjectReference`](https://godoc.org/k8s.io/api/core/v1#ObjectReference) (and possibly should just reuse that type directly?). I think adding/indenting `metadata` is just more verbose and makes it seem like you're about to define a whole object, when it's just a reference.

Re (3), it's currently stuffed into `ref` to allow the possibility to add a sibling field that's the full custom task spec in the future. I agree it looks sort of unnecessary now, hopefully it won't be lonely for long 😄 

Making `name` not required sounds fine by me, I think I need to think/experiment more with a `CELTask` specifically to see how that looks. 🤔